### PR TITLE
Query normalization

### PR DESCRIPTION
**Pull request by camsaul**
Big refactor to "normalize" all queries as they come in, and also whenever they come out of the database in a Card, Metric, or Segment. **For everyone outside of the new `metabase.mbql.normalize` namespace you are guaranteed to be working with a normalized query.** This has several major advantages: i.e

### What's New

*  Manipulating queries is now predictable and simplified 12x. For example, before we had code in a half dozen places to deal with aggregations which could look like any of the following:

   ```clj
   {:aggregation :count}                      ; ancient syntax
   {:aggregation [:count]}                    ; MBQL 98 single aggregation syntax
   {:aggregation [[:count]]}                  ; MBQL 98 multiple aggregations syntax with one ag
   {:aggregation [:count 10]}                 ; single aggregation with a Field
   {:aggregation [[:count 10]]}               ; multiple aggregation syntax with one ag with a Field
   {:aggregation [[:count [:field-id 10]]]}   ; actually the preferred way to write it 
   {:aggregation [["COUNT" ["field_id" 10]]]} ; of course, MBQL is case-insensitive and lisp-case/snake_case insensitive
   ```
   *   Predictably, this made it very hard to manipulate anything at all in the middleware, at least until a query was "expanded". 
   *  There was almost no place where we did manipulate pre-expanded queries that accounted for all of the crazy possibilities that were technically legal.  
   *  When manipulating "expanded" queries you had to add other expanded things, which was a headache in and of itself. And in some other cases it was difficult if not impossible to figure out where to put middleware because it needed some info from an expanded query but needed to take place before the query was expanded and things were resolved
   *  As a proof of how much simpler this makes things I rewrote the `expand-macros` middleware and cut the size in half. See for yourself: [before](https://github.com/metabase/metabase/blob/master/src/metabase/query_processor/middleware/expand_macros.clj) [after](https://github.com/metabase/metabase/blob/648bcfd5637ae219c65bedac51d8a846af66b6ac/src/metabase/query_processor/middleware/expand_macros.clj)
*  Since the frontend will no longer have to deal with all the wacky syntax possibilities we can simplify the code there a lot and remove  helper functions like `mbqlEq`. I noticed this was really only used in a fraction of the places where MBQL tokens were compared, which should eliminate a whole class of possible bugs if we generated queries on the backend that didn't line up with FE expectations (e.g. X-Ray stuff)
*  Since all clauses look the same and all Field IDs are wrapped by `[:field-id ...]` clauses it's trivial to walk a query and see if it refers to a Field. I predict this would be useful for something like column-level permissions. This has already been put to good use in the `expand-macros` middleware rewrite as I was able to optimize it to fetch all Segments/Metrics in one DB call instead of one DB call per Segment. e.g. a query with 3 Metrics resolves them all in 1 DB call now instead of 3. Performance improvements!
*  It is now actually possible to update MBQL syntax without having to rewrite tons of code to support both the old and new versions. Instead, just update things to support only the new version and add logic to transform the old version in `metabase.mbql.normalize`
*  Also trivial to do query "planning" optimizations like replacing `[:not [:not <filter>]]` with `<filter>` or  eliminating duplicate filter clauses (we're already doing a lot of this in `normalize`)
*  Actually the craziest/worst part about the old code was that *all* middleware was expected to handle "expanded" queries, even if the middleware normally ran before query expansion, because some tests were written with `ql/` functions, which produced expanded queries. I actually rewrote every single test that used the old `ql/` functions to use normal queries instead in preparation for removing expanded queries entirely (see below) with the exception of one namespace. (Part of the reason this query is so big)

**TL;DR** Drastic code simplification, a whole class of bugs eliminated, performance improvements, now actually easy to add new features/tweaks to MBQL

### Other highlights of this PR include:

* For the first time ever we have a [complete schema for a valid MBQL query (post-normalization)](https://github.com/metabase/metabase/blob/648bcfd5637ae219c65bedac51d8a846af66b6ac/src/metabase/mbql/schema.clj) using a nice little DSL I slapped together that can be used to validate transformations (e.g., make sure your middleware that added a filter left the query in a valid state). Before there actually wasn't one source of truth as to what MBQL really *was* -- the wiki was missing a few things or incorrect about others; the frontend had a different idea of what was valid from the backend, etc. Previously the only way to tell if a query was valid if it worked when you ran it 😢.
*  A new library, [`metabase.mbql.util`](https://github.com/metabase/metabase/blob/648bcfd5637ae219c65bedac51d8a846af66b6ac/src/metabase/mbql/util.clj), with functions for doing query transformations like `add-filter-clause`. Combined with the schema we can validate all the transformations, and it also does sophisticated stuff like removing duplicate clauses from the query. Of course, this will also help us catch a lot of bugs and hopefully speed up development time 
*  I wrote the new `metabase.mbql` stuff in a way that kept it as a completely separate module from either `metabase.query-processor` or `metabase.models`. It could theoretically be spun off into its own own library. Either way since query-related stuff is done all over the codebase this felt appropriate.

**TL;DR** For the first time ever, a complete grammar for MBQL, new library for doing common query manipulations that validates what it does

### Next Steps

*  Pull a few things like `template-tags` into the MBQL schema and add use the complete schema to validate the results of `normalize` and other middleware steps
*  Now that we don't rely on the `expand` middleware to parse the query and deal with different syntax possibilities, and un-expanded queries are actually in a predictable format, we can remove the query "expansion" altogether. This alone should make our lives 10x easier, especially when we don't have to deal with variants like `FieldPlaceholder`. I'm planning on reworking the `resolve` middleware to simply fetch everything that's needed (Fields, etc.) and keep it in an `info` dict or something and then it can be accessed as needed. Driver query processing code can just dispatch off of MBQL clause name keywords since they'll all be normalized
   *  Yes, this means we can delete all of the record types and schemas in `metabase.query-processor.interface`. I've already marked them all as deprecated. So we won't have an entirely separate record type for Fields and Tables with different key names, we can just use the normal Toucan models. This should reduce cognitive load a bit, especially for contributors, and tear out a solid few kLOC.
   *  Using the new util functions for walking a query and replacing clauses should make it easier for drivers that need to handle certain clauses in special ways. 

### Merging this PR

*  I got everything working except and tested everything out on a running instance except for 4 frontend tests where removing a breakout isn't removing the corresponding `order-by` clause. I thought I fixed `cleanQuery` but it doesn't seem to be working. @tlrobinson I updated the frontend to use the MBQL 98 `[:asc/:desc <field>]` syntax for order by clauses instead of the old `[<field> :ascending/:descending]`. Everything else works correctly otherwise. I couldn't figure out how to fix those last 2 tests. If you could point out what I need to fix that would be amazing 🙏

### CODE REVIEW

**Comment by MarkRx** [comment]
Will this help in being able to mature the UI query builder to do more? For example more than 1 hop - #1173

File: src/metabase/automagic_dashboards/comparison.clj [hunk 1]
```diff
@@ -149,25 +155,24 @@
...
   (get-in card [:visualization_settings :graph.series_labels]
           (map (comp capitalize-first metric-name)
-               (qp.util/get-in-normalized card [:dataset_query :query :aggregation]))))
+               (get-in card [:dataset_query :query :aggregation]))))
 
 (defn- unroll-multiseries
   [card]
   (if (and (multiseries? card)
-           (-> card :display qp.util/normalize-token (= :line)))
+           (-> card :display mbql.u/normalize-token (= :line)))
```
Comment (sbelak) [inline src/metabase/automagic_dashboards/comparison.clj]: why does this need to still be normalized?

File: src/metabase/automagic_dashboards/core.clj [hunk 2]
```diff
@@ -496,25 +504,28 @@
...
                                         (->> context :source u/get-id (str "card__")))}
-                 (not-empty filters)
-                 (assoc :filter (->> filters
-                                     (map :filter)
-                                     (apply qp.expand/merge-filter-clauses)))
-
-                 (not-empty dimensions)
+                 (seq filters)
+                 (assoc :filter (vec
+                                 (cons
```
Comment (sbelak) [inline src/metabase/automagic_dashboards/core.clj]: wouldn't (apply vector :and) be clearer instead of relying on cons prepending to lists?

File: src/metabase/automagic_dashboards/core.clj [hunk 3]
```diff
@@ -496,25 +504,28 @@
...
+                                  :and
+                                  (map (comp (partial normalize/normalize-fragment [:query :filter])
+                                             :filter)
+                                       filters))))
+
+                 (seq dimensions)
                  (assoc :breakout dimensions)
 
-                 (not-empty metrics)
```
Comment (sbelak) [inline src/metabase/automagic_dashboards/core.clj]: is there a reason beyond aesthetic preference to prefer seq to not-empty?

File: src/metabase/automagic_dashboards/core.clj [hunk 4]
```diff
@@ -1138,7 +1149,7 @@
 (defn cell-title
   "Return a cell title given a root object and a cell query."
   [root cell-query]
-  (str/join " " [(if-let [aggregation (qp.util/get-in-normalized (-> root :entity) [:dataset_query :query :aggregation])]
+  (str/join " " [(if-let [aggregation (get-in (-> root :entity) [:dataset_query :query :aggregation])]
```
Comment (sbelak) [inline src/metabase/automagic_dashboards/core.clj]: :entity could be moved into the keys vector

File: src/metabase/automagic_dashboards/core.clj [hunk 5]
```diff
@@ -206,18 +211,21 @@
...
-  (let [[aggregation & _] (qp.util/get-in-normalized card-or-question [:dataset_query :query :aggregation])]
-    (or (nil? aggregation)
-        (and (or (string? aggregation)
-                 (keyword? aggregation))
-             (= (qp.util/normalize-token aggregation) :rows)))))
+  (nil? (get-in card-or-question [:dataset_query :query :aggregation])))
+
+(defn- table-id
+  "For some reason that's not really clear things like the 'Cards' used in this namespace don't have consistent Table ID
+  keys..."
```
Comment (sbelak) [inline src/metabase/automagic_dashboards/core.clj]: This was to support both Cards loaded from the DB as well as those that came from a roundtrip to the FE. If my understanding is correct, we now don't munge keys when sending to the FE, so I think it should be sufficient to just use `table-id` everywhere.

File: src/metabase/automagic_dashboards/filters.clj [hunk 6]
```diff
@@ -175,28 +176,42 @@
...
-         (apply merge-filter-clauses refinement))))
+  (let [in-refinement?  (into #{}
+                              (map collect-field-references)
+                              (flatten-filter-clause refinement))
+        existing-filters (->> filter-clause
+                             flatten-filter-clause
+                             (remove (comp in-refinement? collect-field-references)))]
+    (if (seq existing-filters)
+      ;; for some reason existing-filters and refinement are more often than not non-normalized for reasons I don't
+      ;; understand.
```
Comment (sbelak) [inline src/metabase/automagic_dashboards/filters.clj]: I think it's because in a large part (if xraying anything aside from cards or ad-hoc questions) they are programatically generated and have at this point never passed through normalization.

**Review by sbelak — APPROVED** [review]
Stoked about the changes. This is going to make so many things nicer/cleaner. 

As for CR: just minor nitpickings from my side. 👍

File: src/metabase/automagic_dashboards/core.clj (hunk 3 above)
Comment (camsaul) [reply src/metabase/automagic_dashboards/core.clj]: Just aesthetic. Sorry for changing this. Sometimes I get carried away changing everything. I'll change it back

File: src/metabase/automagic_dashboards/core.clj (hunk 4 above)
Comment (camsaul) [reply src/metabase/automagic_dashboards/core.clj]: oh yeah I'll fix that

File: src/metabase/automagic_dashboards/core.clj (hunk 5 above)
Comment (camsaul) [reply src/metabase/automagic_dashboards/core.clj]: The thing is cards from the DB still have a `table_id` key. I'm not normalizing things like that (although we certainly could normalize all the DB keys; that's a different story). I'll update the docstring

File: src/metabase/automagic_dashboards/core.clj (hunk 3 above)
Comment (sbelak) [reply src/metabase/automagic_dashboards/core.clj]: No worries. I'm happy if we make a coding style decision and unify such things. Clearly I'm in favour of not-empty as it's more semantic, but I'm even more in favour of a unified style.

File: src/metabase/automagic_dashboards/core.clj (hunk 5 above)
Comment (camsaul) [reply src/metabase/automagic_dashboards/core.clj]: I think it might be the things you call "Questions" (?) have `table-id`. Things coming back from a round trip from the FE should still have `table_id`

File: src/metabase/automagic_dashboards/core.clj (hunk 2 above)
Comment (camsaul) [reply src/metabase/automagic_dashboards/core.clj]: Good point

File: src/metabase/automagic_dashboards/comparison.clj (hunk 1 above)
Comment (camsaul) [reply src/metabase/automagic_dashboards/comparison.clj]: Only the `dataset_query` is getting normalized, other properties of the Card still don't get normalized. `:display` is a `:keyword` Toucan type so we shouldn't *need* to normalize this (AFAIK the frontend always uses all-lower-case `"line"`, never `"LINE"`, but I wasn't trying to rewrite more code than I needed to. I broke enough things as it was. So I left this in here since it wouldn't hurt anything

Comment (camsaul) [reply src/metabase/automagic_dashboards/comparison.clj]: I will check with tom and see if this can be removed

File: src/metabase/automagic_dashboards/core.clj (hunk 5 above)
Comment (sbelak) [reply src/metabase/automagic_dashboards/core.clj]: With "question" I mean an adhoc query. In terms of x-ray code these come back from the FE and get stuffed into a `QueryInstance` via models/query/adhoc-query

File: src/metabase/automagic_dashboards/core.clj (hunk 3 above)
Comment (camsaul) [reply src/metabase/automagic_dashboards/core.clj]: Yeah I was going by this 

https://github.com/bbatsov/clojure-style-guide#nil-punning

File: src/metabase/automagic_dashboards/core.clj (hunk 5 above)
Comment (camsaul) [reply src/metabase/automagic_dashboards/core.clj]: ok got it.

File: src/metabase/automagic_dashboards/filters.clj (hunk 6 above)
Comment (camsaul) [reply src/metabase/automagic_dashboards/filters.clj]: thanks. I will update the comment.

Comment (sbelak) [reply src/metabase/automagic_dashboards/filters.clj]: If my understanding is correct and queries attached to cards get normalized on load, it would make sense to add normalization to `models/query/adhoc-query` as well.

**Comment by camsaul** [comment]
@MarkRx yes, it will make adding new features like that easier to do in the future.

File: src/metabase/mbql/util.clj [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+    ;; look for :field-id clauses
+    (clause-instances :field-id {:query {:filter [:= [:field-id 10] 20]}})
+    ;;-> [[:field-id 10]]
+
+    ;; look for :+ or :- clauses
+    (clause-instances #{:+ :-} ...)"
+  {:style/indent 1}
+  [k-or-ks x]
+  (let [instances (transient [])]
```
Comment (senior) [inline src/metabase/mbql/util.clj]: I think you want an atom here, from the [transients reference](https://clojure.org/reference/transients)
```
Note in particular that transients are not designed to be bashed in-place. You must capture and use the return value in the next call.
```

File: src/metabase/automagic_dashboards/core.clj [hunk 8]
```diff
@@ -841,7 +855,7 @@
                                      (encode-base64-json
                                       {:database (:database root)
                                        :type     :query
-                                       :query    {:source_table (->> root
+                                       :query    {:source-table (->> root
```
Comment (senior) [inline src/metabase/automagic_dashboards/core.clj]: How does this PR change where/when/how `snake_case` gets converted to `kabob-case`? Does the FE now use `kabob-case`?

File: src/metabase/mbql/normalize.clj [hunk 9]
```diff
@@ -0,0 +1,541 @@
...
+  #### REMOVING EMPTY CLAUSES
+
+  Removing empty clauses like `{:aggregation nil}` or `{:breakout []}`.
+
+  Token normalization occurs first, followed by canonicalization, followed by removing empty clauses."
+  (:require [clojure.walk :as walk]
+            [medley.core :as m]
+            [metabase.mbql.util :as mbql.u]
+            [metabase.util :as u]
+            [puppetlabs.i18n.core :refer [tru]]))
```
Comment (senior) [inline src/metabase/mbql/normalize.clj]: `metabase.util.i18n` might not have existed when you started writing this, but would be good to get it switched to that.

**Review by senior — APPROVED** [review]
:+1: This is a great step toward a simpler and more flexible query processor!

File: src/metabase/automagic_dashboards/core.clj (hunk 8 above)
Comment (camsaul) [reply src/metabase/automagic_dashboards/core.clj]: Yes, `Card.dataset_query` and other query stuff gets normalized when it comes out of the database. The FE uses lisp-case now since it's guaranteed to be in that normalized format

Comment (camsaul) [reply src/metabase/automagic_dashboards/core.clj]: Only queries are affected, everything else is the same

File: src/metabase/mbql/normalize.clj (hunk 9 above)
Comment (camsaul) [reply src/metabase/mbql/normalize.clj]: yeah I noticed there are still several namespaces that use the puppet i18n namespace I wasn't sure what the rule was when to use which

File: src/metabase/mbql/util.clj (hunk 7 above)
Comment (camsaul) [reply src/metabase/mbql/util.clj]: TIL. I will fix
